File: frontend/src/services/authService.ts
import api from './api';
import type { AuthResponse, LoginRequest, RegisterRequest } from '@/types';

const authService = {
  register: async (userData: RegisterRequest): Promise<AuthResponse> => {
    const response = await api.post<AuthResponse>('/auth/register', userData);
    if (response.data.token) {
      localStorage.setItem('token', response.data.token);
      localStorage.setItem('user', JSON.stringify(response.data));
    }
    return response.data;
  },

  login: async (credentials: LoginRequest): Promise<AuthResponse> => {
    const response = await api.post<AuthResponse>('/auth/login', credentials);
    if (response.data.token) {
      localStorage.setItem('token', response.data.token);
      localStorage.setItem('user', JSON.stringify(response.data));
    }
    return response.data;
  },

  logout: () => {
    localStorage.removeItem('token');
    localStorage.removeItem('user');
  },

  getCurrentUser: (): AuthResponse | null => {
    const userStr = localStorage.getItem('user');
    if (userStr) return JSON.parse(userStr);
    return null;
  },

  isAuthenticated: (): boolean => {
    const token = localStorage.getItem('token');
    return !!token;
  },

  updateProfile: async (data: { name?: string; idNumber?: string }): Promise<{
    id: number;
    username: string;
    email: string;
    role: string;
    name: string | null;
    idNumber: string | null;
    isActive: boolean;
    createdAt: string;
  }> => {
    const response = await api.put('/auth/profile', data);
    const currentUser = authService.getCurrentUser();
    if (currentUser && response.data) {
      const updatedUser = {
        ...currentUser,
        name: response.data.name,
        idNumber: response.data.idNumber,
      };
      localStorage.setItem('user', JSON.stringify(updatedUser));
    }
    return response.data;
  },
};

export default authService;

File: backend/src/main/java/com/dcisman/service/AuthService.java
package com.dcisman.service;

import com.dcisman.dto.AuthResponse;
import com.dcisman.dto.LoginRequest;
import com.dcisman.dto.RegisterRequest;
import com.dcisman.dto.UpdateProfileRequest;
import com.dcisman.entity.User;
import com.dcisman.exception.BadRequestException;
import com.dcisman.repository.UserRepository;
import com.dcisman.security.JwtUtil;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
@Slf4j
public class AuthService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtUtil jwtUtil;

    @Transactional
    public AuthResponse register(RegisterRequest request) {
        log.info("Attempting to register user: {}", request.getUsername());

        // Check if username already exists
        if (userRepository.existsByUsername(request.getUsername())) {
            throw new BadRequestException("Username is already taken");
        }

        // Check if email already exists
        if (userRepository.existsByEmail(request.getEmail())) {
            throw new BadRequestException("Email is already registered");
        }

        // Create new user
        User user = User.builder()
                .username(request.getUsername())
                .email(request.getEmail())
                .passwordHash(passwordEncoder.encode(request.getPassword()))
                .role(User.Role.USER)
                .isActive(true)
                .build();

        User savedUser = userRepository.save(user);
        log.info("User registered successfully: {}", savedUser.getUsername());

        // Generate JWT token
        String token = jwtUtil.generateToken(savedUser.getUsername());

        return AuthResponse.builder()
                .token(token)
                .type("Bearer")
                .id(savedUser.getId())
                .username(savedUser.getUsername())
                .email(savedUser.getEmail())
                .role(savedUser.getRole().name())
                .name(savedUser.getName())
                .idNumber(savedUser.getIdNumber())
                .build();
    }

    public AuthResponse login(LoginRequest request) {
        log.info("Attempting to login user: {}", request.getUsernameOrEmail());

        // Find user by username or email
        User user = userRepository.findByUsername(request.getUsernameOrEmail())
                .orElseGet(() -> userRepository.findByEmail(request.getUsernameOrEmail())
                        .orElseThrow(() -> new BadCredentialsException("Invalid username/email or password")));

        // Check if user is active
        if (!user.getIsActive()) {
            throw new BadRequestException("Account is inactive. Please contact support.");
        }

        // Verify password
        if (!passwordEncoder.matches(request.getPassword(), user.getPasswordHash())) {
            throw new BadCredentialsException("Invalid username/email or password");
        }

        log.info("User logged in successfully: {}", user.getUsername());

        // Generate JWT token
        String token = jwtUtil.generateToken(user.getUsername());

        return AuthResponse.builder()
                .token(token)
                .type("Bearer")
                .id(user.getId())
                .username(user.getUsername())
                .email(user.getEmail())
                .role(user.getRole().name())
                .name(user.getName())
                .idNumber(user.getIdNumber())
                .build();
    }

    public User getCurrentUser(String username) {
        return userRepository.findByUsername(username)
                .orElseThrow(() -> new BadRequestException("User not found: " + username));
    }

    @Transactional
    public User updateProfile(String username, UpdateProfileRequest request) {
        log.info("Attempting to update profile for user: {}", username);

        User user = userRepository.findByUsername(username)
                .orElseThrow(() -> new BadRequestException("User not found: " + username));

        // Update name if provided
        if (request.getName() != null) {
            user.setName(request.getName().trim().isEmpty() ? null : request.getName().trim());
        }

        // Update ID number if provided
        if (request.getIdNumber() != null) {
            user.setIdNumber(request.getIdNumber().trim().isEmpty() ? null : request.getIdNumber().trim());
        }

        User updatedUser = userRepository.save(user);
        log.info("Profile updated successfully for user: {}", username);

        return updatedUser;
    }
}
